lesson5/hometask_TetyanaKarpova.rb

```diff
@@ -0,0 +1,36 @@
+class LazyHash < Hash
+  def lazy?(key)
+    fetch(key).is_a? Proc
+  end
+
+  def []=(key, value)
+    unless self.has_key? key
+      define_singleton_method(key) do
+        value.is_a?(Proc) ? self[key] = value.call : value
+      end
+    end
+    super
+  end
+
+  def self.[](*params)
```

Comment:
Hm. It works, but with issues:
1) You call `#super` twice: every time this method is performed. Makes sense to store result and return it, without duplicated performing. And take a look on [#tap](http://ruby-doc.org/core-2.3.3/Object.html#method-i-tap) method, it can be helpful here.
2) You define methods based on key list in exact lazyhash object for **all instances of `LazyHash` class**. It's incorrect. See,
```ruby
h = LazyHash[:a, proc {"!"}, :b, 2]
h.b # => 2
h2 = LazyHash.new
h2.b # => 2
```
User `#define_singleton_method` instead, you can explicitly specify a receiver:
```ruby
  hash = super
  hash.define_singleton_method ...
```
3. Current implementation of the method does not consider that value could be changed.
```ruby
h = LazyHash[:a, proc {"!"}, :b, 2]
h.b # => 2
h[:b] = 3
h.b # => 2
```
It happens because your dynamically specified methods work only with first `value`, which was received when methods were defined. See the condition in `#[]=`. It checks whether we have `key` in a hash or not. If it presents already, the dynamic method won't be redefined to work with a new `value`. So, even that code would work unexpectedly:
```ruby
h2 = LazyHash.new
h2[:test] = 1
h2.test # => 1
h2[:test] = 2
h2.test # => 1
```

I can propose two possible solutions:
- do not use `self.has_key?` condition in `#[]=` to redefine method forcibly.
- calculate `value` in `define_singleton_method` block by key, not to use initially predefined `value`.

You can apply one of these approaches or implement yours.